### PR TITLE
Docker support

### PR DESCRIPTION
**Pull request by MarcoBuster**
Hi! I'd like to add Docker support for this project. 
I think we'll need to support the passage of configuration via environment variables and add some documentation. Let me know if you like this idea :)

### CODE REVIEW

**Comment by krateng** [comment]
Much appreciated, thanks!

Be aware however that the master branch of the repository was on an old version, and the current version of Maloja is distributed as a Python package. I've merged the new version into the master branch now, so you might want to adjust accordingly.

**Comment by MarcoBuster** [comment]
Thank you!

**Comment by kinduff** [comment]
Any chance this can be merged and deployed automatically to [Dockerhub](https://hub.docker.com/)? I can help out if you want.

**Comment by krateng** [comment]
Alright, since we haven't heard back from @MarcoBuster I've just tried to update it myself. If docker works as I think it works, this should do without having to download anything else but the Dockerfile. @kinduff could you take a look at it or potentially test it?

**Comment by kinduff** [comment]
@krateng Is there a way to bypass the config prompt when starting up the server? Maybe pass the data as environment variables or mount a config file somewhere.

**Comment by krateng** [comment]
Forgot about that, sorry. Should now be possible to load configuration from environment variables with the prefix `MALOJA_`.

**Comment by kinduff** [comment]
@krateng thank you so much for this effort, let me try again and I'll let you know how it goes.

**Comment by MarcoBuster** [comment]
Sorry I stopped answering. @krateng, thank you! I was doing the exact same thing about the load configuration.

**Comment by MarcoBuster** [comment]
> Any chance this can be merged and deployed automatically to [Dockerhub](https://hub.docker.com/)? I can help out if you want.

@kinduff  We can do this with GitHub actions or services like Travis.

**Comment by telyn** [comment]
docker hub also has built-in automated builds from github - so that might be the easiest of all solutions. A few clicks and it's set up, once the PR is merged

**Comment by kinduff** [comment]
Hello, sorry for the delay on my feedback. I agree can do DockerHub + Github instead of Travis or even Github actions, I've seen that work flawlessly and it's intuitive.

In the other hand, I've been trying to modify the current code to make it work as a Docker image. Since the current server de-attaches itself as a processwhenever its ran, we might need to add an option to avoid that. Does this make sense?

**Comment by krateng** [comment]
I've added the command `maloja run` that keeps the main process attached.

**Comment by kinduff** [comment]
@krateng so far so good, I modified the image just to be able to run it, the intention is to be able to move the config folder (currently at `.local/shared/maloja`) as a volume, but I need to make this work first. So here's the `Dockerfile` modified:

```docker
FROM python:3.6-alpine

RUN apk update
RUN apk add gcc libxml2-dev libxslt-dev py3-pip libc-dev

RUN pip3 install malojaserver

ENV HOME /root

WORKDIR $HOME

RUN mkdir -p $HOME/.local/share/maloja
RUN mkdir $HOME/.local/share/maloja/cache
RUN mkdir $HOME/.local/share/maloja/clients
RUN mkdir $HOME/.local/share/maloja/logs
RUN mkdir $HOME/.local/share/maloja/rules
RUN mkdir $HOME/.local/share/maloja/scrobbles
RUN mkdir $HOME/.local/share/maloja/settings

EXPOSE 42010

ENTRYPOINT maloja run
```

Here's how I'm building it:

```shell
docker build . -t maloja/maloja
```

And here's how I'm trying to run it:

```
docker run -p 42010:42010 -e MALOJA_SKIP_SETUP=true -e MALOJA_CACHE_EXPIRE_POSITIVE=300 -e MALOJA_CACHE_EXPIRE_NEGATIVE=30 maloja/maloja
```

Had to add manually the `MALOJA_CACHE_EXPIRE_POSITIVE` and `MALOJA_CACHE_EXPIRE_NEGATIVE` because it was failing due to being unable to parse the integer. I may be missing something, maybe related to the file `settings.ini` or something like that?

Here's the output when I'm trying to run the command above:

```shell
Using PYHP version 2
[maloja.database] Starting database...
[maloja.database] Building database...
[maloja.database] Database fully built!
[maloja.database] Authenticated Machines: 
[maloja.database] Database reachable!
[maloja.server] Starting up Maloja server...
Traceback (most recent call last):
  File "/usr/local/bin/maloja", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/doreah/control.py", line 44, in wrapper
    return func(*args,**kwargs)
  File "/usr/local/lib/python3.6/site-packages/maloja/controller.py", line 195, in main
    if action in actions: actions[action](*args,**kwargs)
  File "/usr/local/lib/python3.6/site-packages/maloja/controller.py", line 169, in direct
    from . import server
  File "/usr/local/lib/python3.6/site-packages/maloja/server.py", line 282, in <module>
    waitress.serve(webserver, host=HOST, port=MAIN_PORT, threads=THREADS)
  File "/usr/local/lib/python3.6/site-packages/waitress/__init__.py", line 12, in serve
    server = _server(app, **kw)
  File "/usr/local/lib/python3.6/site-packages/waitress/server.py", line 52, in create_server
    adj = Adjustments(**kw)
  File "/usr/local/lib/python3.6/site-packages/waitress/adjustments.py", line 317, in __init__
    setattr(self, k, self._param_map[k](v))
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

I was not able to continue debugging a cause.

**Comment by krateng** [comment]
First of all, thank you for taking the time and effort!

I forgot that some of the functionality relies on a default settings file being generated first, which doesn't happen when we run the server directly. I'll try to rework the structure a bit over the weekend so that these things can be setup first.

**Comment by kinduff** [comment]
@krateng No need to thanks, it's a pleasure. Sounds good, I'll try to make it work with the current setup! Going to import the `settings.ini` and also modify the image to build it directly from the source instead of pulling from pip. Let's keep it flowing!

**Comment by krateng** [comment]
With the latest update (v2.3.6 on PyPI) it should now work without any of the CACHE_EXPIRE variables being set. You should also be able to skip the mkdir commands.

**Comment by FoxxMD** [comment]
Looking forward to this <3

**Comment by FoxxMD** [comment]
@krateng I forked @MarcoBuster 's branch and rebased it with master. His dockerfile works without modification.I used @kinduff 's build command without the cache stuff and added a host environmental variable because the ipv6 default for host wasn't working for me.

```
docker build -t maloja . && docker run -p 42010:42010 --env MALOJA_SKIP_SETUP=true --env MALOJA_HOST=0.0.0.0 maloja 
```

that worked perfectly 👍

**Comment by krateng** [comment]
Fantastic, I'll merge it then.
